lib/eventsource/index.js

```diff
@@ -0,0 +1,448 @@
+  }
+
+  /**
+   * Returns a boolean indicating whether the EventSource object was
+   * instantiated with CORS credentials set (true), or not (false, the default).
+   */
+  get withCredentials () {
+    return this.#withCredentials
+  }
+
+  async #connect () {
+    if (this.#readyState === CLOSED) return
+
+    this.#readyState = CONNECTING
+
+    const fetchParam = {
+      request: this.#request
+    }
+
+    // 14. Let processEventSourceEndOfBody given response res be the following step: if res is not a network error, then reestablish the connection.
+    const processEventSourceEndOfBody = (response) => {
+      if (isNetworkError(response)) {
+        this.dispatchEvent(new ErrorEvent('error', { error: response.error }))
+        this.close()
+      }
+
+      this.#connect()
+    }
+
+    // 15. Fetch request, with processResponseEndOfBody set to processEventSourceEndOfBody...
+    fetchParam.processResponseEndOfBody = processEventSourceEndOfBody
+
+    // and processResponse set to the following steps given response res:
+
+    fetchParam.processResponse = (response) => {
+      // 1. If res is an aborted network error, then fail the connection.
+
+      if (isNetworkError(response)) {
+        // 1. When a user agent is to fail the connection, the user agent
+        // must queue a task which, if the readyState attribute is set to a
+        // value other than CLOSED, sets the readyState attribute to CLOSED
+        // and fires an event named error at the EventSource object. Once the
+        // user agent has failed the connection, it does not attempt to
+        // reconnect.
+        if (response.aborted) {
+          this.close()
+          this.dispatchEvent(new ErrorEvent('error', { error: response.error }))
+          return
+          // 2. Otherwise, if res is a network error, then reestablish the
+          // connection, unless the user agent knows that to be futile, in
+          // which case the user agent may fail the connection.
+        } else {
+          this.#reconnect()
+          return
+        }
+        // 3. Otherwise, if res's status is not 200, [...], then fail the
+        // connection.
+      } else if (response.status !== 200) {
+        let message = `Unexpected status code: ${response.status}`
+
+        if (response.status === 204) {
+          message = 'No content'
+        }
+        this.close()
+        this.dispatchEvent(new ErrorEvent('error', { message }))
+        return
+      }
+
+      // 3. Otherwise, [...] if res's
+      // `Content-Type` is not `text/event-stream`, then fail the
+      // connection.
+
+      const contentType = response.headersList.get('content-type', true)
+      const mimeType = contentType !== null ? parseMIMEType(contentType) : 'failure'
+      if (mimeType === 'failure' || mimeType.essence !== 'text/event-stream') {
+        this.close()
+        this.dispatchEvent(new ErrorEvent('error', { message: 'Invalid content-type' }))
+        return
+      }
```

Comment:
It would probably be easier to read if the ifs were merged and the step not split in two

Reply:
I had the feeling that it has less complexity than putting everything into one block. Will investigate into this

Reply:
I could implement it like this:
Do you like it better?

```js
      // 3. Otherwise, if res's status is not 200, or if res's `Content-Type`
      // is not `text/event-stream`, then fail the connection.
      const contentType = response.headersList.get('content-type', true)
      const mimeType = contentType !== null ? parseMIMEType(contentType) : 'failure'
      const contentTypeValid = mimeType !== 'failure' && mimeType.essence === 'text/event-stream'
      if (
        response.status !== 200 ||
        contentTypeValid === false
      ) {
        let message

        if (response.status !== 200) {
          message = 'No content'
        } else if (contentTypeValid === false) {
          message = `Invalid content-type: ${contentType}`
        } else {
          message = `Invalid status code: ${response.status}`
        }

        this.close()
        this.dispatchEvent(new ErrorEvent('error', { message }))
        return
      }

```

Reply:
yes, I prefer this way (specifically the comment, makes it easier to search for)